### PR TITLE
feat: Resolve and install

### PR DESCRIPTION
**Pull request by Aradhya-Tripathi**
### Dependency resolver

* Added `--resolve-deps` to get-app (get-app ++).
   * **Does not install dependencies of an app unless ```--resolve-deps``` flag is passed.**
   * Option first resolves all dependencies of the frappe app passed as an argument.
   * Dependencies read from hooks.py.
   * **Does not clone app repo before resolving** all of its dependencies and their dependencies.
   * After obtaining the resolution all the dependencies are installed.

* Added app states 
   * Added `app_states.json` to maintain app states i.e app versions, remotes, resolution, etc.

### CODE REVIEW

**Review by gavindsouza — COMMENTED** [review]
After #1208, get-app resolves dependencies mentioned in required_apps and there's no turn off switch either. I realised that could be problematic given the resolution is still naive. Thought this PR would change that.

Ran the following two commands to check the difference between them. There was none. Both initialized bench with frappe@develop, erpnext@develop & healthcare@develop. 

* `bench --use-feature='dynamic-feed' get-app healthcare --init-bench`
* `bench --use-feature='dynamic-feed' get-app healthcare --init-bench --resolve-deps`

This behaviour makes sense though. With `--init-bench`, dependencies should always get resolved IMO. Else, your install is likely going to break.

---

But, when using get-app just to install apps on a particular bench. I would expect a difference between using `--resolve-deps` and not...In this case, there's not? 😅

Screenshot 2022-02-16 at 12 47 44 PM

**Comment by Aradhya-Tripathi** [comment]
> After #1208, get-app resolves dependencies mentioned in required_apps and there's no turn off switch either. I realised that could be problematic given the resolution is still naive. Thought this PR would change that.
> 
> Ran the following two commands to check the difference between them. There was none. Both initialized bench with frappe@develop, erpnext@develop & healthcare@develop.
> 
> * `bench --use-feature='dynamic-feed' get-app healthcare --init-bench`
> * `bench --use-feature='dynamic-feed' get-app healthcare --init-bench --resolve-deps`
> 
> This behaviour makes sense though. With `--init-bench`, dependencies should always get resolved IMO. Else, your install is likely going to break.
> 
> But, when using get-app just to install apps on a particular bench. I would expect a difference between using `--resolve-deps` and not...In this case, there's not? sweat_smile
> 
> Screenshot 2022-02-16 at 12 47 44 PM

It would make more sense to use --resolve-deps once versioning is introduced, as of now the difference is internal, resolution is made internally before installation starts.

We could probably log the resolution or display it during --resolve-deps.

File: bench/app.py [hunk 1]
```diff
@@ -171,42 +170,80 @@ def get(self):
...
+	@step(title="Cloning and installing {repo}", success="App {repo} Installed")
+	def install_resolved_apps(self, *args, **kwargs):
+		self.get()
+		self.install(*args, **kwargs, resolved=True)
+
 	@step(title="Uninstalling App {repo}", success="App {repo} Uninstalled")
 	def uninstall(self):
 		self.bench.run(f"{self.bench.python} -m pip uninstall -y {self.repo}")
 
+	def _get_dependencies(self):
```
Comment (gavindsouza) [inline bench/app.py:218]: 1. Handle `on_disk`
2. Fix this

```python
>>> a = App("./apps/healthcare")
>>> a._get_dependencies()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/gavin/bench/bench/app.py", line 216, in _get_dependencies
    required_apps = eval(lines[0].strip("required_apps").strip().lstrip("=").strip())
IndexError: list index out of range
>>>
```

File: bench/app.py [hunk 2]
```diff
@@ -310,16 +318,28 @@ def get_app(
...
 
 	bench = Bench(bench_path)
 	app = App(git_url, branch=branch, bench=bench)
 	git_url = app.url
 	repo_name = app.repo
 	branch = app.tag
 	bench_setup = False
+	frappe_path, frappe_branch = None, None
+
+	if resolve_deps:
```
Comment (gavindsouza) [inline bench/app.py:346]: Make behaviour consistent for command with and without `--resolve-deps`

File: bench/app.py [hunk 3]
```diff
@@ -371,22 +405,75 @@ def get_app(
...
+	for repo_name, app in reversed(resolution.items()):
+		existing_dir, path_to_app = check_existing_dir(bench_path, repo_name)
+		if existing_dir:
+			is_compatible = False
+
+			try:
+				installed_branch = bench.apps.states[repo_name]["resolution"]["branch"].strip()
+			except:
+				installed_branch = (
+					subprocess.
```
Comment (gavindsouza) [inline bench/app.py:442]: Check scope for using the `gitpython` library here

File: bench/utils/app.py [hunk 4]
```diff
@@ -107,7 +107,9 @@ def switch_to_develop(apps=None, bench_path=".", upgrade=True):
 
 
 def get_version_from_string(contents, field="__version__"):
-	match = re.search(r"^(\s*%s\s*=\s*['\\\"])(.+?)(['\"])(?sm)" % field, contents)
+	match = re.search(r"^(\s*%s\s*=\s*['\\\"])(.+?)(['\"])" % field, contents, flags=(re.S | re.M))
+	if not match:
+		raise Exception("No match was found")
```
Comment (gavindsouza) [inline bench/utils/app.py]: Throw named exceptions - also sonarcloud

File: bench/utils/__init__.py [hunk 5]
```diff
@@ -48,6 +49,37 @@ def is_frappe_app(directory: str) -> bool:
...
+	:param frappe_branch: branch to check
+	:type frappe_branch: str
+	:raises InvalidRemoteException: branch for this repo doesn't exist
+	"""
+	if "http" in frappe_path and frappe_branch:
+		frappe_path = frappe_path.replace(".git", "")
+		try:
+			owner, repo = frappe_path.split("/")[3], frappe_path.split("/")[4]
+		except IndexError:
+			raise InvalidRemoteException
```
Comment (gavindsouza) [inline bench/utils/__init__.py]: Add a message for this? Seems confusing when hit with tracebacks (ref: https://github.com/frappe/bench/issues/1267)

File: bench/app.py [hunk 6]
```diff
@@ -171,42 +170,80 @@ def get(self):
...
-		setup_app_dependencies(
-			repo_name=self.repo,
-			bench_path=self.bench.name,
-			branch=self.tag,
-			verbose=verbose,
-			skip_assets=skip_assets,
-		)
+		if not resolved and self.repo != "frappe":
+			click.secho(
+				f"Ignoring dependencies of {self.name} to install dependencies use --resolve-deps",
```
Comment (gavindsouza) [inline bench/app.py]: Handle this bit with the setup requirements command too. Perhaps add a `--resolve-deps` option too and don't show this message when called from there anyway

**Review by gavindsouza — CHANGES_REQUESTED** [review]

File: bench/app.py (hunk 2 above)
Comment (Aradhya-Tripathi) [reply bench/app.py:346]: This is just to provide support for frappe versions (when that's mentioned in the hooks of the required apps).

File: bench/app.py (hunk 6 above)
Comment (Aradhya-Tripathi) [reply bench/app.py]: Not showing the message when called from setup requirements.

File: bench/utils/__init__.py (hunk 5 above)
Comment (Aradhya-Tripathi) [reply bench/utils/__init__.py]: Resolved in [cd583f7](https://github.com/frappe/bench/pull/1257/commits/cd583f7822a5aaa29f09cc9ba016b6b23c1d9cc2)

File: bench/utils/app.py (hunk 4 above)
Comment (Aradhya-Tripathi) [reply bench/utils/app.py]: Resolved in 
a32ceb4

File: bench/app.py (hunk 1 above)
Comment (Aradhya-Tripathi) [reply bench/app.py:218]: fixed by [0b862c0](https://github.com/frappe/bench/pull/1257/commits/0b862c07457c9914840bd0ce87a90306fec37456)

File: bench/app.py (hunk 3 above)
Comment (Aradhya-Tripathi) [reply bench/app.py:442]: Git gives a lot more flexibility with ease so I think it would be better to use git directly here?

**Review by gavindsouza — APPROVED** [review]
Works as expected, LGTM

**Pull request merged**